feature-gate/program/src/processor.rs

```diff
@@ -0,0 +1,89 @@
+//! Program state processor
+
+use {
+    crate::{error::FeatureGateError, instruction::FeatureGateInstruction},
+    borsh::BorshDeserialize,
+    solana_program::{
+        account_info::{next_account_info, AccountInfo},
+        entrypoint::ProgramResult,
+        feature::Feature,
+        msg,
+        program::invoke,
+        program_error::ProgramError,
+        pubkey::Pubkey,
+        system_instruction, system_program,
+    },
+};
+
+/// Processes an [Activate](enum.FeatureGateInstruction.html) instruction.
+pub fn process_activate(program_id: &Pubkey, accounts: &[AccountInfo]) -> ProgramResult {
+    let account_info_iter = &mut accounts.iter();
+
+    let feature_info = next_account_info(account_info_iter)?;
+    let authority_info = next_account_info(account_info_iter)?;
+    let _system_program_info = next_account_info(account_info_iter)?;
+
+    if !feature_info.is_signer || !authority_info.is_signer {
+        return Err(ProgramError::MissingRequiredSignature);
+    }
+
+    if feature_info.owner != &system_program::id() {
+        return Err(FeatureGateError::FeatureNotSystemAccount.into());
+    }
+
+    invoke(
+        &system_instruction::allocate(feature_info.key, Feature::size_of() as u64),
+        &[feature_info.clone()],
+    )?;
+    invoke(
+        &system_instruction::assign(feature_info.key, program_id),
+        &[feature_info.clone()],
+    )?;
+
+    Ok(())
+}
+
+/// Processes an [revoke](enum.FeatureGateInstruction.html) instruction.
+pub fn process_revoke(program_id: &Pubkey, accounts: &[AccountInfo]) -> ProgramResult {
+    let account_info_iter = &mut accounts.iter();
+
+    let feature_info = next_account_info(account_info_iter)?;
+    let destination_info = next_account_info(account_info_iter)?;
+    let authority_info = next_account_info(account_info_iter)?;
+
+    if !authority_info.is_signer {
+        return Err(ProgramError::MissingRequiredSignature);
+    }
```

Comment:
This currently allows *anyone* to revoke a feature as long as you have any random signature aftewards, which could be pretty catastrophic :sweat_smile: 

Since activation requires a signature from the feature id keypair, let's require a signature from the feature id keypair for revoke, and not worry about some other authority. What do you think?

Unless I'm missing something -- have there been discussions around an extra authority?

Reply:
Correct and nope, I honestly have no idea how the current authority setup works, that's kind where I'm leaving off on this thing until we see how the talks go

Reply:
Feature keypair for revoke makes sense, since the contributor is instructed to hold onto this keypair right?

Reply:
Do we require an authority to activate? I guess to my understanding, anyone can make a feature account and mark it owned by `Feature111111111111111111111111111111111111 `, but if it's not in the [feature set](https://github.com/solana-labs/solana/blob/f4287d70bb78307f16949509e013fb5e6f15a751/sdk/src/feature_set.rs) it doesn't get picked up (?)

Reply:
Yeah that's exactly it. So it's all done by the feature keypair. I think for v1 of this program, we keep the same paradigm